xblock/fields.py

```diff
@@ -454,14 +460,16 @@ def _check_or_enforce_type(self, value):
                 value, traceback.format_exc().splitlines()[-1])
             warnings.warn(message, FailingEnforceTypeWarning, stacklevel=3)
         else:
-            try:
-                equal = value == new_value
-            except TypeError:
-                equal = False
-            if not equal:
-                message = "The value {!r} would be enforced to {!r}".format(
-                    value, new_value)
-                warnings.warn(message, ModifyingEnforceTypeWarning, stacklevel=3)
+            # Refuse to evaluate lazy evaluations during this type enforcement.
+            if not self._is_lazy(value) and not self._is_lazy(new_value):
```

Comment:
Would it make more sense to check instead if one of the values is lazy, and if so coerce both to unicode text before doing the comparison?  That would just move the condition check inside the try block, leaving the rest of the method unchanged from before.  Is there a problem with the default value being evaluated before the i18n machinery is set up?

Reply:
Exactly - this code avoids the evaluation of the lazily translated values before Django's i18n functionality is ready. Coercing or comparing them when one or both is lazily translated always ended up evaluating and causing this error:
```
======================================================================
1) ERROR: xblock.test.django.test_field_translation.test_db_model_keys
----------------------------------------------------------------------
   Traceback (most recent call last):
    .tox/py27-django111/lib/python2.7/site-packages/nose/case.py line 197 in runTest
      self.test(*self.arg)
    xblock/test/django/test_field_translation.py line 49 in test_db_model_keys
      assert_equals(tester.str_field, None)
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/functional.py line 148 in __eq__
      return self.__cast() == other
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/functional.py line 131 in __cast
      return self.__text_cast()
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/functional.py line 119 in __text_cast
      return func(*self.__args, **self.__kw)
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/translation/__init__.py line 89 in ugettext
      return _trans.ugettext(message)
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/translation/trans_real.py line 345 in ugettext
      return do_translate(message, 'ugettext')
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/translation/trans_real.py line 321 in do_translate
      _default = _default or translation(settings.LANGUAGE_CODE)
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/translation/trans_real.py line 228 in translation
      _translations[language] = DjangoTranslation(language)
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/translation/trans_real.py line 129 in __init__
      self._add_installed_apps_translations()
    .tox/py27-django111/lib/python2.7/site-packages/django/utils/translation/trans_real.py line 171 in _add_installed_apps_translations
      "The translation infrastructure cannot be initialized before the "
   AppRegistryNotReady: The translation infrastructure cannot be initialized before the apps registry is ready. Check that you don't make non-lazy gettext calls at import time.
```